test/lib/test_utils.hh

```diff
@@ -61,4 +61,18 @@ extern boost::test_tools::assertion_result has_scylla_test_env(boost::unit_test:
 future<bool> compare_files(std::string fa, std::string fb);
 future<> touch_file(std::string name);
 
+extern std::mutex boost_logger_mutex;
+
+#define THREADSAFE_BOOST_CHECK( BOOST_CHECK_EXPR ) {                    \
+        std::lock_guard<std::mutex> guard(tests::boost_logger_mutex);   \
+        BOOST_CHECK_EXPR;                                               \
```

Comment:
please note, this lock not only serializes the mutations to the `unit_test_log` instance, it also serializes the evaluations of `BOOST_CHECK_EXPR`. so **all** expressions passed to `BOOST_REQUIRE*` and `BOOST_CHECK*` (checkpoints for short) are serialized, this substantially changes the way how the tests are executed:

before this change, the expressions passed to checkpoints are performed in parallel.

after this change, they are serialized at every checkpoint defined by the newly introduced `THREADSAFE_BOOST_*` macros. if we want to verify the behavior of the interact of different shards in a multi-shard environment, we would sacrifice some test cases because of the serialization of the check points.

one way to address this, is to, only serialize the mutations to the `unit_test_log` instance, by

1. extract the expression with side effects out of the macro, so it won't be evaluated *in* the serialized tasks
2. redefine macros like `BOOST_TEST_TOOL_ET_IMPL` and `BOOST_TEST_TOOL_ET_IMPL_EX` if we are using the new macros like `BOOST_TEST_REQUIRE` or redefine macros like `BOOST_TEST_TOOL_IMPL` if we are using the old-style macros like `BOOST_REQUIRE`, to serialize its body to, for instance, shard 0.

i understand this is non-trivial in comparison to the grand-lock solution proposed in this PR.

but it is a design decision we need to make. it depends on what we expect from the tests. and i think we need to make this explicit.

Reply:
The checks are usually side-effect-free, and normally adding a lock to a seastar application doesn't affect (improve) its correctness, only decreases the speed.

Since we're only using it in one place, I think we can live with the less-than-perfect implementation.